### ISSUE TITLE
TypeScript 2.6 does not work with XrmQuery

### ISSUE THREAD

**Issue by henrikhannemose** [delegateas/XrmDefinitelyTyped#33]
When using TypeScript 2.6, there appears to be an issue with (at least) retrieveMultiple.

Running the [example of retrieveMultiple from the wiki](https://github.com/delegateas/XrmDefinitelyTyped/wiki/XrmQuery-Web-API#usage), we get an error since we cannot filter on an empty object:

![RetrieveMultiple, empty object](https://user-images.githubusercontent.com/1955579/32546427-5d08430c-c47f-11e7-8323-506e28c2b2a7.png)

The entire definition for retrieveMultiple appears to be broken (`XrmQuery.retrieveMultiple<Account_Select, {}, {}, {}, {}, {}>(...)`)
![XrmQuery.retrieveMultiple](https://user-images.githubusercontent.com/1955579/32546501-a0c0dd98-c47f-11e7-9f4d-adab6c65dce4.png)

What has changed with TypeScript 2.6 (I have not tested with TypeScript 2.5 to see if that works or not), and what can we do to make XrmQuery work?

**Comment by zaha** [comment]
I experience the same issue using Typescript 2.6.1

**Comment by TomMalow** [comment]
I have a potential solution.
XDT generates an interface `WebEntities` for each entity. This interface holds a single intersection type for the interface which is an intersection of:
- `WebMappingRetrieve`
- `WebMappingCUD`
- `WebMappingRelated`

However, we are never using more than one of these type in our method calls. The `retrieve` / `retrieveMultiple` call always uses `WebEntities` as a `WebMappingRetrieve` interface. The update call uses it as the `WebMappingCUD` interface. So there are no reason why `WebEntities` should have an intersection type other than reduce to the amount of interfaces that the definition holds.
My proposal is then to change
```
interface WebEntities {
  accounts: WebMappingRetrieve<Account_Select,Account_Expand,Account_Filter,Account_Fixed,Account_Result,Account_FormattedResult> & WebMappingCUD<Account_Create,Account_Update> & WebMappingRelated<Account_RelatedOne,Account_RelatedMany>;
}
```
to the following
```
interface WebEntitiesRetrieve {
  accounts: WebMappingRetrieve<Account_Select,Account_Expand,Account_Filter,Account_Fixed,Account_Result,Account_FormattedResult>;
}

interface WebEntitiesRelated {
    accounts: WebMappingRelated<Account_RelatedOne, Account_RelatedMany>;
}

interface WebEntitiesCUD {
    accounts: WebMappingCUD<Account_Create, Account_Update>;
}
```
for each entity.
The method calls then needs to be changed to expect one of the three interfaces. So the retrieve:
```
function retrieve<ISelect, IExpand, IFixed, FormattedResult, Result>(entityPicker: (x: WebEntities) => WebMappingRetrieve<ISelect, IExpand, any, IFixed, Result, FormattedResult>, id: string): XQW.RetrieveRecord<ISelect, IExpand, IFixed, FormattedResult, Result>;
```
Would be changed to
```
function retrieve<ISelect, IExpand, IFixed, FormattedResult, Result>(entityPicker: (x: WebEntitiesRetrieve) => WebMappingRetrieve<ISelect, IExpand, any, IFixed, Result, FormattedResult>, id: string): XQW.RetrieveRecord<ISelect, IExpand, IFixed, FormattedResult, Result>;
```

This proposal all hangs on my assumption that the `WebEntities` interface is always reduced down to on of the three types in all of the method calls. But from my limited testing it seems to work.

I think, and this is just a guess, that the previous code was "incorrect" from the start but TypeScript simply did not have proper checking on intersection types. This was then fixed in 2.6
